enhancements/cluster-logging/content-filter.md

```diff
@@ -0,0 +1,261 @@
+---
+title: content-filter
+authors:
+  - "@alanconway"
+reviewers: 
+  - "@jcantrill"
+approvers:
+  - "@jcantrill"
+api-approvers: 
+  - "@jcantrill"
+creation-date: 2023-11-03
+last-updated:  2023-11-03
+tracking-link:
+  - [LOG-2155 Filter log messages based on metadata and content.](https://issues.redhat.com/browse/LOG-2155)
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Content filters for log records
+
+## Summary
+
+Allow users to reduce the volume of log data by:
+1. Dropping unwanted log records completely.
+2. Pruning unwanted fields from log records.
+
+## Motivation
+
+Collecting all logs from a cluster produces a large amount of data, which can be expensive to transport and store.
+A lot of log data is low-value noise that does not need to be stored.
+
+The ideal solution is to configure each application to generate only valuable log data, however:
+- Not all applications can be tuned to remove noise while keeping important data.
+- Configuration of applications is not always under the control of the cluster administrator.
+- The value of logs varies depending on use: for example debugging a specific problem vs. monitoring overall health.
+
+### User Stories
+
+* As a logging adminstrator I want to drop log records by severity or other field values in the record.
+* As a logging administrator I want to prune log records by removing pod annotations, or other uninteresting fields.
+
+### Goals
+
+The user can configure the ClusterLogForwarder to drop and prune log records according to their requirements.
+
+### Non-Goals
+
+No complex record transformations, for example adding new fields with computed values.
+
+## Proposal
+
+### Workflow Description
+
+**log administrator** is a human responsible for configuring log collection, storage and forwarding in a cluster.
+
+1. The log administrator creates named `filter` sections in a `ClusterLogForwarder`
+2. Pipelines that require filtering have a `filterRef` field with a list of filter names.
+3. The collector edits forwarded log streams according to the filters.
+
+### API Extensions
+
+A new `filters` section in the `ClusterLogForwarder` allows named filters to be defined.
+This proposal defines two types of fitler "prune" and "drop".
+
+#### Prune filters
+
+Remove all listed fields from each record.
+
+##### API
+
+``` yaml
+spec:
+  filters:
+  - name:          #user-selected name
+    type: prune
+    prune:
+      remove: []         #remove all fields on this list.
+      removeOthers: []   #remove all fields not on this list.
```

Comment:
@jcantrill sounds like:
  `remove` is a blacklist. Fields to be deleted
   `removeOthers` is a white list. Fields to maintain

Initially, it was sounding bad to me the name of the variable, but it seems good after a second review

Reply:
@r2d2rnd correct, and clearly the names are not very clear.  Any suggestions? I thought "remove/keep" but since "keep" really means "remove everything that is not in the keep-list" so worried that wasn't clear.

Reply:
To me, the definition of the word "prune" is to remove or drop leaving all others alone.  This implies there is no need for a black/white list; just provide the list of fields to drop.  As an example,  you "prune" branches from trees, selectively removing those you do not want.  The fact you left branches alone is implied:

```
the process of removing dead, diseased or damaged branches, as well as shaping and thinning out the plant
```
 
IMO, it would be more intentional to provide a counter filter of "keep" which means: only keep **this** list fields and remove all others

Reply:
What bugs me is both black- and white-lists have the effect of removing fields, so both are kinds of "pruning". 
How about this ; re-use the terminology of the "drop" rule (dunno why I didn't in the first place) to give: 
```
prune:
  - in: [ remove-if-listed ] 
  - notIn: [ remove-if-not-listed ]
```  
Also means we could (but don't have to) extend pruning to allow regexps or other types of match in a way that is consistent between both types of rules.

Reply:
I'm proposing the filter drops the notion of includes or exclude; you just provide a list of fields.  The name "prune" or "keep" implies what is happening.  Additionally you could chain filters to achieve include/exclude